src/main/java/software/amazon/msk/auth/iam/internals/IAMSaslClient.java

```diff
@@ -203,28 +204,58 @@ private static boolean isChallengeEmpty(byte[] challenge) {
         return true;
     }
 
+    public static class ClassLoaderAwareIAMSaslClientFactory implements SaslClientFactory {
+
+        @Override
+        public SaslClient createSaslClient(String[] mechanisms,
+                String authorizationId,
+                String protocol,
+                String serverName,
+                Map<String, ?> props,
+                CallbackHandler cbh) throws SaslException {
+            String mechanismName = getMechanismNameForClassLoader(cbh.getClass().getClassLoader());
+
+            // Creat a client by delegating to the SaslClientFactory for the classloader of the CallbackHandler
+            return Sasl.createSaslClient(
+                    new String[] { mechanismName },
+                    authorizationId, protocol, serverName, props, cbh);
+        }
+
+        @Override
+        public String[] getMechanismNames(Map<String, ?> props) {
+            return new String[] { IAMLoginModule.MECHANISM };
+        }
+    }
+
     public static class IAMSaslClientFactory implements SaslClientFactory {
+
         @Override
         public SaslClient createSaslClient(String[] mechanisms,
                 String authorizationId,
                 String protocol,
                 String serverName,
                 Map<String, ?> props,
                 CallbackHandler cbh) throws SaslException {
+            String mechanismName = getMechanismNameForClassLoader(getClass().getClassLoader());
+
             for (String mechanism : mechanisms) {
-                if (IAMLoginModule.MECHANISM.equals(mechanism)) {
+                if (mechanismName.equals(mechanism)) {
                     return new IAMSaslClient(mechanism, cbh, serverName, new AWS4SignedPayloadGenerator());
                 }
             }
+
             throw new SaslException(
-                    "Requested mechanisms " + Arrays.asList(mechanisms) + " not supported. The supported" +
-                            "mechanism is " + IAMLoginModule.MECHANISM);
+                    "Requested mechanisms " + Arrays.asList(mechanisms) + " not supported. " +
+                            "The supported mechanism is " + mechanismName);
         }
 
         @Override
         public String[] getMechanismNames(Map<String, ?> props) {
-            return new String[] { IAMLoginModule.MECHANISM };
+            return new String[] { getMechanismNameForClassLoader(getClass().getClassLoader()) };
         }
     }
 
+    public static String getMechanismNameForClassLoader(ClassLoader classLoader) {
```

Comment:
If you do create a `IAMClassAwareLoginModule`, this static method could be a utility static method in that class.